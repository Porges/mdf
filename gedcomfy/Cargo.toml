[package]
name = "gedcomfy"
version = "0.1.0"
edition.workspace = true
license.workspace = true
authors.workspace = true

[features]
default = ["miette-highlighting", "kdl"]
miette-highlighting = ["miette/fancy-no-backtrace"]
legacy-encodings = ["dep:oem_cp"]
kdl = ["dep:kdl"]

[dependencies]
ascii = "1.1.0"
encoding_rs = "0.8.34"
# TODO: update once this is released https://github.com/zkat/miette/pull/393
miette = { git = "https://github.com/porges/miette", features = ["fancy-no-backtrace"], branch = "inherit-source-to-causes" }
oem_cp = { version = "2.0.0", optional = true }
owo-colors = { version = "4.0.0", features = ["supports-colors"] }
paste = "1.0.15"
thiserror = "1.0.63"
tracing = { version = "0.1", features = ["attributes"] }
vec1 = "1.12.1"
kdl = { version = "4.6.0", optional = true }
yoke = { version = "0.7.4", features = ["derive"] }

[dev-dependencies]
insta = { version = "1.39.0", features = ["glob"] }
