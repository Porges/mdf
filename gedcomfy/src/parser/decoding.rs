use std::borrow::Cow;

use crate::{
    parser::encodings::external_file_encoding,
    versions::{parse_version_head_gedc_vers, GEDCOMVersion},
    FileStructureError,
};

use super::{
    encodings::{DetectedEncoding, EncodingError, InvalidDataForEncodingError},
    lines::LineSyntaxError,
    records::{read_first_record, RawRecord, RecordStructureError},
    versions::VersionError,
    GEDCOMSource, Sourced,
};

/// Attempts to detect the encoding of a GEDCOM file and provide the data
/// in a decoded format, so that it can be parsed.
///
/// ## Details
///
/// Ah, _encoding_.
///
/// GEDCOM is a classic file format which has a chicken-and-egg problem:
/// the encoding of the file is specified in the file itself, but the file
/// cannot be read without knowing the encoding.
///
/// This function attempts to discover the encoding of a GEDCOM file by
/// several methods which it will attempt in order:
///
/// 1. Firstly, we try to detect the encoding ‘externally’, without parsing the GEDCOM
///    records:
///
///    a. We check for a Byte Order Mark (BOM) at the start of the file. If one of
///       these is found (for UTF-8, or UTF-16 BE/LE), it most likely can be trusted.
///
///    b. Otherwise, it will try to determine the encoding by content-sniffing the first
///       character in the file, which should always be a literal '0'. (The start of a
///       legitimate file must always begin with `0 HEAD <newline>`). This can determine
///       some non-ASCII-compatible encodings such as UTF-16.
///    
///    If one of those methods work, we then parse the file to double-check the encoding
///    is correct, and that the encoding agrees with what is specified in the file,
///    and extract the file version.
///
/// 2. Otherwise, if there is no BOM and the encoding is something that is ASCII-compatible,
///    we must parse the records to determine the encoding. In order to do this,
///    the file is parsed in a minimally-decoding mode which only decodes the record
///    levels and tag names (which both must consist of characters in the ASCII subset).
///    
///    The further tricky thing here is that different versions of the GEDCOM standard
///    specify the encoding differently. In version 5 files, the encoding is specified
///    in the `GEDC.VERS` record, while in version 7 files, the `GEDC.VERS` record is
///    not permitted and files _must_ be encoded in UTF-8. So, if we knew the version
///    up-front, we could determine the encoding from that. Instead, we must discover
///    it by—guess what—parsing the file.
///
/// 3. If neither of those methods work, the file assumed to not be GEDCOM file, and
///    an error indicating this is returned.
///
/// If you want to exert more control about how the version or encoding are determined,
/// you can pass appropriate options to the [`parse`] function. See the documentation
/// on [`detect_file_encoding_opt`].
pub fn detect_and_decode(input: &[u8]) -> Result<(GEDCOMVersion, Cow<str>), DecodingError> {
    if let Some(external_encoding) = external_file_encoding(input)? {
        // now we can decode the file to actually look inside it
        let decoded = external_encoding.decode(input)?;
        // get version and double-check encoding with file
        let ext_enc = external_encoding.encoding;
        let (version, f_enc) = parse_gedcom_header(decoded.as_ref(), Some(external_encoding))?;
        // we don’t need the encoding here since we already decoded
        // it will always be the same
        debug_assert_eq!(f_enc.encoding, ext_enc);
        Ok((*version, decoded))
    } else {
        // we need to determine the encoding from the file itself
        let (version, file_encoding) = parse_gedcom_header(input, None)?;
        // now we can actually decode the input
        let decoded = file_encoding.decode(input)?;
        Ok((*version, decoded))
    }
}

#[derive(thiserror::Error, Debug, miette::Diagnostic)]
pub enum DecodingError {
    #[error("Unable to determine version of GEDCOM file")]
    #[diagnostic(transparent)]
    VersionError(#[from] VersionError),

    #[error("Unable to determine encoding of GEDCOM file")]
    #[diagnostic(transparent)]
    EncodingError(#[from] EncodingError),

    #[error("GEDCOM file contained data which was invalid in the detected encoding")]
    #[diagnostic(transparent)]
    InvalidDataForEncoding(#[from] InvalidDataForEncodingError),

    #[error("GEDCOM file structure is invalid")]
    #[diagnostic(transparent)]
    FileStructureError(#[from] FileStructureError),

    #[error("GEDCOM file contains a record-hierarchy error")]
    #[diagnostic(transparent)]
    RecordStructureError(#[from] RecordStructureError),

    #[error("GEDCOM file contains a syntax error")]
    #[diagnostic(transparent)]
    SyntaxError(#[from] LineSyntaxError),
}

pub fn parse_gedcom_header<S: GEDCOMSource + ?Sized>(
    input: &S,
    external_encoding: Option<DetectedEncoding>,
) -> Result<(Sourced<GEDCOMVersion>, DetectedEncoding), DecodingError> {
    let first_record = read_first_record::<_, DecodingError>(input)?;
    let head = first_record
        .as_ref()
        .and_then(|r| r.ensure_tag("HEAD"))
        .ok_or_else(|| FileStructureError::MissingHeadRecord {
            span: first_record.as_ref().map(|r| r.span),
        })?;

    let version = detect_version_from_head_record(head)?;
    let encoding = version.detect_encoding_from_head_record(head, external_encoding)?;
    Ok((version, encoding))
}

fn detect_version_from_head_record<S: GEDCOMSource + ?Sized>(
    head: &Sourced<RawRecord<S>>,
) -> Result<Sourced<GEDCOMVersion>, VersionError> {
    if let Some(gedc) = head.subrecord_optional("GEDC") {
        tracing::debug!("located GEDC record");
        if let Some(vers) = gedc.subrecord_optional("VERS") {
            tracing::debug!("located VERS record");
            // GEDCOM 4.x or above (including 5.x and 7.x)
            let data = vers.line.data.expect("TODO: error");
            return data
                .try_map(|d| parse_version_head_gedc_vers(d))
                .map_err(|source| VersionError::InvalidVersion {
                    source,
                    span: data.span,
                });
        }
    }

    if let Some(sour) = head.subrecord_optional("SOUR") {
        // GEDCOM 2.x or 3.0
        if let Some(vers) = sour.subrecord_optional("VERS") {
            // this is 3.0 – TODO check line data value
            return Ok(Sourced {
                value: GEDCOMVersion::V3,
                span: vers.line.span,
            });
        }

        todo!("2.x handling")
    }

    Err(VersionError::NoVersion {})
}
