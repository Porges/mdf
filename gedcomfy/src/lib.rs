//! This is a library for parsing and validating GEDCOM files.

use core::str;
use std::borrow::Cow;

use encodings::{DataError, GEDCOMEncoding, MissingRequiredSubrecord};
use miette::SourceSpan;
use parser::{
    decoding::DecodingError,
    lines::{iterate_lines, LineSyntaxError},
    options::{OptionSetting, ParseOptions},
    records::RawRecord,
    GEDCOMSource, Sourced,
};
use vec1::Vec1;
use versions::GEDCOMVersion;

pub mod encodings;
pub mod highlighting;
mod ntypes;
pub mod parser;
pub(crate) mod v5;
pub(crate) mod v7;
pub(crate) mod versions;

#[derive(thiserror::Error, Debug, miette::Diagnostic)]
pub enum ValidationError {
    #[error("{} syntax error{} detected", .errors.len(), if .errors.len() > 1 { "s" } else { "" })]
    SyntaxErrorsDetected {
        #[related]
        errors: Vec1<LineSyntaxError>,
    },
    #[error("Encoding error detected: further validation errors will not be found")]
    #[diagnostic(transparent)]
    EncodingErrorDetected {
        #[from]
        error: DecodingError,
    },
}

impl<'a, S: GEDCOMSource + ?Sized> RawRecord<'a, S> {
    pub(crate) fn subrecord_optional(&self, subrecord_tag: &str) -> Option<&Sourced<RawRecord<S>>> {
        self.records
            .iter()
            .find(|r| r.value.line.tag.value == subrecord_tag)
    }

    pub(crate) fn subrecord(
        &self,
        subrecord_tag: &'static str,
        subrecord_description: &'static str,
    ) -> Result<&Sourced<RawRecord<S>>, MissingRequiredSubrecord> {
        self.subrecord_optional(subrecord_tag)
            .ok_or(MissingRequiredSubrecord {
                tag: subrecord_tag,
                description: subrecord_description,
            })
    }

    pub(crate) fn subrecords_optional(
        &self,
        tag: &'static str,
    ) -> impl Iterator<Item = &Sourced<RawRecord<S>>> {
        self.records
            .iter()
            .filter(move |r| r.value.line.tag.value == tag)
    }

    pub(crate) fn subrecords_required(
        &self,
        tag: &'static str,
        description: &'static str,
    ) -> Result<Vec1<&Sourced<RawRecord<S>>>, MissingRequiredSubrecord> {
        let v = Vec::from_iter(self.subrecords_optional(tag));
        Vec1::try_from(v).map_err(|_| MissingRequiredSubrecord { tag, description })
    }
}

/// Checks that the lines in the file are (minimally) well-formed.
/// Returns the number of lines in the file if successful.
pub fn validate_syntax(source: &[u8], buffer: &mut String) -> Result<usize, ValidationError> {
    validate_syntax_opt(source, buffer, ParseOptions::default())
}

pub fn validate_syntax_opt(
    source: &[u8],
    buffer: &mut String,
    parse_options: ParseOptions,
) -> Result<usize, ValidationError> {
    let (_version, source) = parser::decoding::detect_and_decode(source, parse_options)?;
    let source: &str = match source {
        Cow::Borrowed(input) => input,
        Cow::Owned(owned) => {
            *buffer = owned;
            buffer.as_str()
        }
    };

    let mut line_count = 0;
    let errors = Vec::from_iter(iterate_lines(source).filter_map(|r| match r {
        Ok(_) => {
            line_count += 1;
            None
        }
        Err(e) => Some(e),
    }));

    if let Ok(errors) = Vec1::try_from(errors) {
        Err(ValidationError::SyntaxErrorsDetected { errors })
    } else {
        Ok(line_count)
    }
}

pub(crate) struct FileFormatOptions {
    pub(crate) version_option: OptionSetting<GEDCOMVersion>,
    pub(crate) encoding_option: OptionSetting<GEDCOMEncoding>,
}

#[derive(thiserror::Error, Debug, miette::Diagnostic)]
pub(crate) enum FileStructureError {
    #[error("Missing HEAD record")]
    #[diagnostic(code(gedcom::schema_error::missing_head_record))]
    MissingHeadRecord {
        #[label("this is the first record in the file; the HEAD record should appear before it")]
        span: Option<SourceSpan>,
    },

    #[error("Missing trailer (TRLR) record")]
    #[diagnostic(
        code(gedcom::schema_error::missing_trailer_record),
        help("this record is always required at the end of the file â€“ GEDCOM file might be truncated?")
    )]
    MissingTrailerRecord,

    #[error("Records after trailer (TRLR) record")]
    #[diagnostic(
        code(gedcom::schema_error::records_after_trailer),
        help(
            "there are additional records after the trailer record which marks the end of the file"
        )
    )]
    RecordsAfterTrailer {
        #[label("this record appears after the TRLR record")]
        span: SourceSpan,
    },

    #[error(transparent)]
    #[diagnostic(transparent)]
    DataError(#[from] DataError<'static>),
}
