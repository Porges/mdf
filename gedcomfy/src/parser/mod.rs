use std::borrow::Cow;

use ascii::{AsciiChar, AsciiStr};
use decoding::DecodingError;
use lines::iterate_lines;
use miette::{SourceOffset, SourceSpan};
use options::ParseOptions;
use records::{RawRecord, RecordBuilder};

pub mod encodings;
pub mod lines;
pub mod options;
pub mod records;

pub(crate) mod decoding;
pub(crate) mod versions;

/// Represents the minimal amount of decoding needed to
/// parse information from GEDCOM files.
pub trait GEDCOMSource: ascii::AsAsciiStr + PartialEq<AsciiStr> {
    fn lines(&self) -> impl Iterator<Item = &Self>;
    fn split_once(&self, char: AsciiChar) -> Option<(&Self, &Self)>;
    fn split_once_opt(&self, char: AsciiChar) -> (&Self, Option<&Self>) {
        match self.split_once(char) {
            Some((a, b)) => (a, Some(b)),
            None => (self, None),
        }
    }
    fn span_of(&self, source: &Self) -> SourceSpan;
    fn starts_with(&self, char: AsciiChar) -> bool;
    fn ends_with(&self, char: AsciiChar) -> bool;
    fn is_empty(&self) -> bool;
    fn slice_from(&self, offset: usize) -> &Self;
}

impl GEDCOMSource for str {
    fn lines(&self) -> impl Iterator<Item = &Self> {
        // GEDCOM lines are terminated by "any combination of a carriage return and a line feed"
        (*self).split(|c| c == '\r' || c == '\n').map(|mut s| {
            while s.starts_with('\n') || s.starts_with('\r') {
                s = &s[1..];
            }

            s
        })
    }

    fn span_of(&self, source: &Self) -> SourceSpan {
        debug_assert!(source.as_ptr() >= self.as_ptr());
        SourceSpan::new(
            SourceOffset::from(unsafe { source.as_ptr().byte_offset_from(self.as_ptr()) } as usize),
            source.len(),
        )
    }

    fn starts_with(&self, char: AsciiChar) -> bool {
        (*self).starts_with(char.as_char())
    }

    fn ends_with(&self, char: AsciiChar) -> bool {
        (*self).ends_with(char.as_char())
    }

    fn is_empty(&self) -> bool {
        (*self).is_empty()
    }

    fn slice_from(&self, offset: usize) -> &Self {
        &(*self)[offset..]
    }

    fn split_once(&self, char: AsciiChar) -> Option<(&Self, &Self)> {
        (*self).split_once(char.as_char())
    }
}

impl GEDCOMSource for [u8] {
    fn lines(&self) -> impl Iterator<Item = &Self> {
        // GEDCOM lines are terminated by "any combination of a carriage return and a line feed"
        (*self).split(|&x| x == b'\r' || x == b'\n').map(|mut s| {
            while s.starts_with(&[b'\n']) || s.starts_with(&[b'\r']) {
                s = &s[1..];
            }

            s
        })
    }

    fn span_of(&self, source: &Self) -> SourceSpan {
        debug_assert!(source.as_ptr() >= self.as_ptr());
        SourceSpan::new(
            SourceOffset::from(unsafe { source.as_ptr().byte_offset_from(self.as_ptr()) } as usize),
            source.len(),
        )
    }

    fn starts_with(&self, char: AsciiChar) -> bool {
        (*self).starts_with(&[char.as_byte()])
    }

    fn ends_with(&self, char: AsciiChar) -> bool {
        (*self).ends_with(&[char.as_byte()])
    }

    fn is_empty(&self) -> bool {
        (*self).is_empty()
    }

    fn slice_from(&self, offset: usize) -> &Self {
        &(*self)[offset..]
    }

    fn split_once(&self, char: AsciiChar) -> Option<(&Self, &Self)> {
        let ix = self.iter().position(|&x| x == char.as_byte())?;
        let (before, after) = self.split_at(ix);
        Some((before, &after[1..]))
    }
}

/// A value that is sourced from a specific location in a GEDCOM file.
///
/// This is used in many places to ensure that we can track back values
/// to their original location, which means that we can provide good
/// diagnostics in the case of errors.
#[derive(Copy, Clone, PartialEq, Eq, Debug)]
pub struct Sourced<T> {
    pub value: T,
    pub span: SourceSpan,
}

impl<T> Sourced<T> {
    pub(crate) fn as_ref(&self) -> Sourced<&T> {
        Sourced {
            value: &self.value,
            span: self.span,
        }
    }

    pub(crate) fn map<U>(self, f: impl FnOnce(T) -> U) -> Sourced<U> {
        Sourced {
            value: f(self.value),
            span: self.span,
        }
    }

    pub(crate) fn try_map<U, E>(self, f: impl FnOnce(T) -> Result<U, E>) -> Result<Sourced<U>, E> {
        Ok(Sourced {
            value: f(self.value)?,
            span: self.span,
        })
    }

    pub(crate) fn try_into<U>(self) -> Result<Sourced<U>, T::Error>
    where
        T: TryInto<U>,
    {
        match self.value.try_into() {
            Ok(value) => Ok(Sourced {
                value,
                span: self.span,
            }),
            Err(err) => Err(err),
        }
    }
}

impl<T, E> Sourced<Result<T, E>> {
    pub(crate) fn transpose(self) -> Result<Sourced<T>, E> {
        match self.value {
            Ok(value) => Ok(Sourced {
                value,
                span: self.span,
            }),
            Err(err) => Err(err),
        }
    }
}

/// A [`Sourced``] value derefs to the inner value, making
/// it easier to work with when the source information is not needed.
impl<T> std::ops::Deref for Sourced<T> {
    type Target = T;
    fn deref(&self) -> &Self::Target {
        &self.value
    }
}

pub fn parse<'a>(
    input: &'a [u8],
    buffer: &'a mut String,
) -> Result<Vec<Sourced<RawRecord<'a>>>, DecodingError> {
    parse_opt(input, buffer, ParseOptions::default())
}

pub fn parse_opt<'a>(
    input: &'a [u8],
    buffer: &'a mut String,
    parse_options: ParseOptions,
) -> Result<Vec<Sourced<RawRecord<'a>>>, DecodingError> {
    let (_version, decoded_input) = decoding::detect_and_decode(input, parse_options)?;

    // if we want to return records without copying data in most cases,
    // the caller must provide a buffer where we can copy data if needed
    // (this happens when parsing UTF-16, or ANSEL if any non-ASCII characters are present)
    let input: &'a str = match decoded_input {
        Cow::Borrowed(borrowed) => borrowed,
        Cow::Owned(owned) => {
            *buffer = owned;
            buffer.as_str()
        }
    };

    let mut record_builder = RecordBuilder::new();

    let mut result = Vec::new();
    for line in iterate_lines(input) {
        match line {
            Ok(line) => {
                if let Some(record) = record_builder.handle_line(line)? {
                    result.push(record);
                }
            }
            Err(err) => return Err(err.into()),
        }
    }

    if let Some(record) = record_builder.complete()? {
        result.push(record);
    }

    Ok(result)
}
