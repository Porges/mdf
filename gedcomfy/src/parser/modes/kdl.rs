use kdl::{KdlDocument, KdlEntry, KdlNode};

use crate::parser::{
    lines::LineValue, records::RawRecord, AnySourceCode, NonFatalHandler, ParseMode, ResultBuilder,
    Sourced,
};

#[derive(Default)]
pub(in crate::parser) struct Mode {}

impl NonFatalHandler for Mode {
    fn non_fatal<E>(&mut self, _error: E) -> Result<(), E>
    where
        E: Into<crate::parser::ParseError> + miette::Diagnostic,
    {
        Ok(())
    }
}

impl ParseMode for Mode {
    type ResultBuilder<'i> = Builder;

    fn get_result_builder<'i>(
        self,
        _version: crate::versions::SupportedGEDCOMVersion,
        _source_code: AnySourceCode,
    ) -> Result<Self::ResultBuilder<'i>, crate::parser::ParseError> {
        Ok(Builder {
            mode: self,
            doc: KdlDocument::new(),
        })
    }
}

pub(in crate::parser) struct Builder {
    mode: Mode,
    doc: KdlDocument,
}

impl NonFatalHandler for Builder {
    fn non_fatal<E>(&mut self, error: E) -> Result<(), E>
    where
        E: Into<crate::parser::ParseError> + miette::Diagnostic,
    {
        self.mode.non_fatal(error)
    }
}

impl<'i> ResultBuilder<'i> for Builder {
    type Result = KdlDocument;

    fn handle_record(
        &mut self,
        record: Sourced<RawRecord<'i>>,
    ) -> Result<(), crate::parser::ParseError> {
        self.doc.nodes_mut().push(record_to_kdl(record.value));
        Ok(())
    }

    fn complete(self) -> Result<Self::Result, crate::parser::ParseError> {
        Ok(self.doc)
    }
}

fn record_to_kdl(record: RawRecord) -> KdlNode {
    let mut node = KdlNode::new(record.line.tag.to_string());

    if let Some(xref) = &record.line.xref {
        node.entries_mut()
            .push(KdlEntry::new_prop("xref", xref.value.to_string()));
    }

    if let Some(mapped) = match record.line.line_value.value {
        LineValue::Ptr(None) => Some(KdlEntry::new_prop("see", kdl::KdlValue::Null)),
        LineValue::Ptr(Some(value)) => Some(KdlEntry::new_prop("see", value)),
        LineValue::Str(data) => Some(KdlEntry::new(data.to_string())),
        LineValue::None => None,
    } {
        node.entries_mut().push(mapped);
    }

    if record.records.is_empty() {
        return node;
    }

    let mut children = KdlDocument::new();
    for subrecord in record.records {
        children.nodes_mut().push(record_to_kdl(subrecord.value));
    }

    node.set_children(children);
    node
}
