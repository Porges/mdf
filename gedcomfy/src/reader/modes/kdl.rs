use kdl::{KdlDocument, KdlEntry, KdlNode};

use crate::reader::{
    lines::LineValue, records::RawRecord, NonFatalHandler, ReadMode, ResultBuilder, Sourced,
};

#[derive(Default)]
pub(in crate::reader) struct Mode {}

impl NonFatalHandler for Mode {
    fn report<E>(&mut self, _error: E) -> Result<(), E>
    where
        E: Into<crate::reader::ReaderError> + miette::Diagnostic,
    {
        Ok(())
    }
}

impl<'i> ReadMode<'i> for Mode {
    type ResultBuilder = Builder;

    fn into_result_builder(
        self,
        _version: crate::versions::SupportedGEDCOMVersion,
    ) -> Result<Self::ResultBuilder, crate::reader::ReaderError> {
        Ok(Builder {
            mode: self,
            doc: KdlDocument::new(),
        })
    }
}

pub(in crate::reader) struct Builder {
    mode: Mode,
    doc: KdlDocument,
}

impl NonFatalHandler for Builder {
    fn report<E>(&mut self, error: E) -> Result<(), E>
    where
        E: Into<crate::reader::ReaderError> + miette::Diagnostic,
    {
        self.mode.report(error)
    }
}

impl<'i> ResultBuilder<'i> for Builder {
    type Result = KdlDocument;

    fn handle_record(
        &mut self,
        record: Sourced<RawRecord>,
    ) -> Result<(), crate::reader::ReaderError> {
        self.doc
            .nodes_mut()
            .push(record_to_kdl(record.sourced_value));
        Ok(())
    }

    fn complete(self) -> Result<Self::Result, crate::reader::ReaderError> {
        Ok(self.doc)
    }
}

fn record_to_kdl(record: RawRecord) -> KdlNode {
    let mut node = KdlNode::new(record.line.tag.to_string());

    if let Some(xref) = &record.line.xref {
        node.entries_mut()
            .push(KdlEntry::new_prop("xref", xref.sourced_value.to_string()));
    }

    if let Some(mapped) = match record.line.value.sourced_value {
        LineValue::Ptr(None) => Some(KdlEntry::new_prop("see", kdl::KdlValue::Null)),
        LineValue::Ptr(Some(value)) => Some(KdlEntry::new_prop("see", value)),
        LineValue::Str(data) => Some(KdlEntry::new(data.to_string())),
        LineValue::None => None,
    } {
        node.entries_mut().push(mapped);
    }

    if record.records.is_empty() {
        return node;
    }

    let mut children = KdlDocument::new();
    for subrecord in record.records {
        children
            .nodes_mut()
            .push(record_to_kdl(subrecord.sourced_value));
    }

    node.set_children(children);
    node
}
