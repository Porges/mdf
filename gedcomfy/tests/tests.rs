use std::{path::PathBuf, sync::Once};

use gedcomfy::parser::{
    encodings::detect_external_encoding, lines::LineValue, options::ParseOptions, parse,
    records::RawRecord,
};
use kdl::{KdlDocument, KdlEntry, KdlNode};
use miette::{NamedSource, Report};

static INIT: Once = Once::new();
fn ensure_hook() {
    INIT.call_once(|| {
        miette::set_hook(Box::new(|_diag| {
            Box::new(
                miette::MietteHandlerOpts::new()
                    .terminal_links(false)
                    .unicode(true)
                    .color(false)
                    .width(132)
                    .build(),
            )
        }))
        .unwrap();
    });
}

#[test]
fn can_parse_allged_lines() {
    let mut path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    path.push("tests/external/others/allged.ged");

    let data = std::fs::read(path).unwrap();
    let mut buffer = String::new();

    let line_count = gedcomfy::validate_syntax(&data, &mut buffer).unwrap();
    assert_eq!(line_count, 1159);
}

#[test]
fn can_parse_allged_fully() -> miette::Result<()> {
    let mut path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    path.push("tests/external/others/allged.ged");

    let parsed_file = gedcomfy::parse_file(&path, ParseOptions::default())?;
    insta::assert_debug_snapshot!(parsed_file);
    Ok(())
}

#[test]
fn produces_expected_allged_tree() {
    let mut path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    path.push("tests/external/others/allged.ged");

    let data = std::fs::read(path).unwrap();
    let buffer = &mut String::new();
    let parsed = parse(&data, buffer).unwrap();

    insta::assert_snapshot!(to_kdl(parsed.into_iter().map(|r| r.value)));
}

#[test]
fn torture_test_valid() {
    ensure_hook();

    insta::glob!("external/torture-test-55-files/*.ged", |path| {
        let data = std::fs::read(path).unwrap();
        let filename = path.file_name().unwrap().to_string_lossy();
        let buffer = &mut String::new();
        let parsed = parse(&data, buffer)
            .map_err(|e| Report::new(e).with_source_code(NamedSource::new(filename, data.clone())))
            .unwrap();
        insta::assert_snapshot!(to_kdl(parsed.into_iter().map(|r| r.value)));
    });
}

#[test]
fn golden_files() -> miette::Result<()> {
    ensure_hook();

    insta::glob!("syntax_inputs/*.ged", |path| {
        let data = std::fs::read(path).unwrap();
        let filename = path.file_name().unwrap();
        insta::with_settings!({
            // provide GEDCOM source alongside output
            description => String::from_utf8_lossy(&data),
        }, {
            match parse(&data, &mut String::new()) {
                Ok(records) => {
                    let kdl = to_kdl(records.into_iter().map(|r| r.value));
                    insta::assert_snapshot!(kdl);
                }
                Err(err) => insta::assert_snapshot!(format!(
                    "{:?}",
                    Report::new(err).with_source_code(NamedSource::new(filename.to_string_lossy(), data))
                )),
            };
        });
    });

    insta::glob!("format_inputs/*.ged", |path| {
        let data = std::fs::read(path).unwrap();
        let filename = path.file_name().unwrap().to_string_lossy();
        insta::with_settings!({
            // provide GEDCOM source alongside output
            description => String::from_utf8_lossy(&data),
        }, {
            match parse(&data, &mut String::new()) {
                Ok(records) => {
                    let kdl = to_kdl(records.into_iter().map(|r| r.value));
                    insta::assert_snapshot!(kdl);
                }
                Err(err) => insta::assert_snapshot!(format!(
                    "{:?}",
                    Report::new(err).with_source_code(NamedSource::new(filename, data))
                )),
            };
        });
    });

    Ok(())
}

fn to_kdl<'a>(records: impl Iterator<Item = RawRecord<'a>>) -> KdlDocument {
    let mut doc = KdlDocument::new();
    for record in records {
        doc.nodes_mut().push(record_to_kdl(record));
    }

    doc
}

fn record_to_kdl(record: RawRecord) -> KdlNode {
    let mut node = KdlNode::new(record.line.tag.to_string());

    if let Some(xref) = &record.line.xref {
        node.entries_mut()
            .push(KdlEntry::new_prop("xref", xref.value.to_string()));
    }

    if let Some(mapped) = match record.line.line_value.value {
        LineValue::Ptr(None) => Some(KdlEntry::new_prop("see", kdl::KdlValue::Null)),
        LineValue::Ptr(Some(value)) => Some(KdlEntry::new_prop("see", value)),
        LineValue::Str(data) => Some(KdlEntry::new(data.to_string())),
        LineValue::None => None,
    } {
        node.entries_mut().push(mapped);
    }

    if record.records.is_empty() {
        return node;
    }

    let mut children = KdlDocument::new();
    for subrecord in record.records {
        children.nodes_mut().push(record_to_kdl(subrecord.value));
    }

    node.set_children(children);
    node
}

#[test]
fn test_encodings() {
    ensure_hook();

    insta::glob!("encoding_inputs/*.ged", |path| {
        let data = std::fs::read(path).unwrap();
        let data_ref: &[u8] = data.as_ref();
        let filename = path.file_name().unwrap().to_string_lossy();
        let encoding_report = match detect_external_encoding(data_ref) {
            Ok(Some(detected)) => format!(
                "External encoding detected: {}\nReason: {}",
                detected.encoding(),
                Report::new(detected.reason())
                    .with_source_code(NamedSource::new(filename.clone(), data.clone()))
            ),
            Ok(None) => "No external encoding detected (ASCII-compatible)".to_string(),
            Err(err) => format!(
                "{}",
                Report::new(err).with_source_code(NamedSource::new(filename.clone(), data.clone()))
            ),
        };

        insta::with_settings!({
            // provide GEDCOM source alongside output
            description => String::from_utf8_lossy(&data),
        }, {
            insta::assert_snapshot!(encoding_report);
            match parse(&data, &mut String::new()) {
                Ok(records) => {
                    let kdl = to_kdl(records.into_iter().map(|r| r.value));
                    insta::assert_snapshot!(kdl);
                }
                Err(err) => insta::assert_snapshot!(format!(
                    "{:?}",
                    Report::new(err).with_source_code(NamedSource::new(filename, data))
                )),
            };
        });
    });
}
