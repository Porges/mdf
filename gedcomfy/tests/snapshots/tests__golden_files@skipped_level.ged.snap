---
source: gedcomfy/tests/tests.rs
description: "0 HEAD\r\n1 GEDC\r\n2 VERS 5.5.1\r\n1 CHAR ASCII\r\n0 TAG\r\n2 TAG\r\n"
expression: "format!(\"{:?}\", miette::Report::new(err))"
input_file: gedcomfy/tests/syntax_inputs/skipped_level.ged
---
gedcom::record_error::invalid_child_level

  × GEDCOM file contains a record-hierarchy error
  ╰─▶ Invalid child level 2, expected 1 or less
